src/main/java/gregtech/api/capability/impl/FluidTankList.java

```diff
@@ -1,130 +1,146 @@
 package gregtech.api.capability.impl;
 
 import gregtech.api.capability.IMultipleTankHandler;
-import net.minecraft.nbt.NBTBase;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.nbt.NBTTagList;
 import net.minecraftforge.common.util.Constants;
 import net.minecraftforge.common.util.INBTSerializable;
 import net.minecraftforge.fluids.FluidStack;
-import net.minecraftforge.fluids.FluidTank;
 import net.minecraftforge.fluids.IFluidTank;
-import net.minecraftforge.fluids.capability.IFluidHandler;
 import net.minecraftforge.fluids.capability.IFluidTankProperties;
 
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
-import java.util.*;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
 
-/**
- * Recommended to use this with {@link NotifiableFluidTankFromList} to ensure
- * proper behavior of the "allowSameFluidFill" setting, but not required.
- */
-public class FluidTankList implements IFluidHandler, IMultipleTankHandler, INBTSerializable<NBTTagCompound> {
+public class FluidTankList implements IMultipleTankHandler, INBTSerializable<NBTTagCompound> {
 
-    protected final List<IFluidTank> fluidTanks;
-    protected IFluidTankProperties[] properties;
+    private final MultiFluidTankEntry[] fluidTanks;
     private final boolean allowSameFluidFill;
+
     private IFluidTankProperties[] fluidTankProperties;
-    private final int hashCode;
 
     public FluidTankList(boolean allowSameFluidFill, IFluidTank... fluidTanks) {
-        this.fluidTanks = Arrays.asList(fluidTanks);
+        ArrayList<MultiFluidTankEntry> list = new ArrayList<>();
+        for (IFluidTank tank : fluidTanks) list.add(wrapIntoEntry(tank));
+        this.fluidTanks = list.toArray(new MultiFluidTankEntry[0]);
         this.allowSameFluidFill = allowSameFluidFill;
-        this.hashCode = Arrays.hashCode(fluidTanks);
     }
 
-    public FluidTankList(boolean allowSameFluidFill, List<? extends IFluidTank> fluidTanks) {
-        this.fluidTanks = new ArrayList<>(fluidTanks);
+    public FluidTankList(boolean allowSameFluidFill, @Nonnull List<? extends IFluidTank> fluidTanks) {
```

Comment:
```suggestion
    public FluidTankList(boolean allowSameFluidFill, @Nonnull Iterable<IFluidTank> fluidTanks) {
```

Reply:
This constructor was present before this, so changing the signature would risk another compat break.